core/variables.js

```diff
@@ -259,72 +259,182 @@ Blockly.Variables.realizePotentialVar = function(varName, varType, potentialVarW
+          // This check ensures that if a new variable is being created from a
+          // workspace that already has a variable of the same name and type as
+          // a potential variable, that potential variable gets turned into a
+          // real variable and thus there aren't duplicate options in the field_variable
+          // dropdown.
+          if (potentialVarMap && opt_type) {
+            variable = Blockly.Variables.realizePotentialVar(validatedText,
+                opt_type, workspace, false);
+          }
+          if (!variable) {
+            variable = workspace.createVariable(validatedText, opt_type);
+          }
+
+          var flyout = workspace.isFlyout ? workspace : workspace.getFlyout();
+          var variableBlockId = variable.getId();
+          if (flyout.setCheckboxState) {
+            flyout.setCheckboxState(variableBlockId, true);
+          }
+
+          if (opt_callback) {
+            opt_callback(variableBlockId);
           }
         } else {
           // User canceled prompt without a value.
           if (opt_callback) {
             opt_callback(null);
           }
         }
-      }, opt_type, modalTitle);
-  };
-  promptAndCheckWithAlert('');
+      }, modalTitle);
+};
+
+/**
+ * This function provides a common interface for variable name validation agnostic
+ * of type. This is so that functions like Blockly.Variables.createVariable and
+ * Blockly.Variables.renameVariable can call a single function (with a single type
+ * type signature) to validate the user-provided name for a variable.
+ * @param {string} type The type of the variable for which the provided name
+ *     should be validated.
+ * @param {string} text The user-provided text that should be validated as a
+ *     variable name.
+ * @param {!Blockly.Workspace} workspace The workspace on which to validate the
+ *     variable name. This is the workspace used to check whether the variable
+ *     already exists.
+ * @param {function(?string=)=} opt_callback An optional function to be called on
+ *     a pre-existing variable of the user-provided name. This function is currently
+ *     only used for broadcast messages.
+ * @return {string} The validated name according to the parameters given, if
+ *     the name is determined to be valid, or null if the name
+ *     is determined to be invalid/in-use, and the calling function should not
+ *     proceed with creating or renaming the variable.
+ * @private
+ */
+Blockly.Variables.nameValidator_ = function(type, text, workspace, opt_callback) {
+  // The validators for the different variable types require slightly different arguments.
+  // For broadcast messages, if a broadcast message of the provided name already exists,
+  // the validator needs to call a function that updates the selected
+  // field option of the dropdown menu of the block that was used to create the new message.
+  // For scalar variables and lists, the validator needs a retry function that is responsible
```

Comment:
I don't see a retry function.  Is this comment out of date?

Reply:
fixed.